crates/wasmtime/src/limits.rs

```diff
@@ -0,0 +1,256 @@
+use crate::{Store, StoreInner};
+use std::rc::Weak;
+
+pub(crate) const DEFAULT_INSTANCE_LIMIT: usize = 10000;
+pub(crate) const DEFAULT_TABLE_LIMIT: usize = 10000;
+pub(crate) const DEFAULT_MEMORY_LIMIT: usize = 10000;
+
+/// Used by hosts to limit resource consumption of instances at runtime.
+///
+/// A [`Store`] can be created with a resource limiter so that hosts can take into account
+/// non-WebAssembly resource usage to determine if a linear memory or table should grow.
+pub trait ResourceLimiter {
+    /// Notifies the resource limiter that an instance's linear memory has been requested to grow.
+    ///
+    /// * `current` is the current size of the linear memory in WebAssembly page units.
+    /// * `desired` is the desired size of the linear memory in WebAssembly page units.
+    /// * `maximum` is either the linear memory's maximum or a maximum from an instance allocator,
+    ///   also in WebAssembly page units. A value of `None` indicates that the linear memory is
+    ///   unbounded.
+    ///
+    /// This function should return `true` to indicate that the growing operation is permitted or
+    /// `false` if not permitted.
+    ///
+    /// Note that this function will be called even when the desired count exceeds the given maximum.
+    ///
+    /// Returning `true` when a maximum has been exceeded will have no effect as the linear memory
+    /// will not be grown.
+    fn memory_growing(
+        &self,
+        store: &Store,
```

Comment:
One alternative design perhaps is that the `Store` isn't passed as an argument here and users manually `RefCell` and `Weak` themselves with the `Store`. Do you think, though, that this will be commonly desired such that we may as well go ahead and expose it?

Reply:
Okay, getting back to this PR: I guess we don't have to expose the `Store` and let them capture whatever context they want since it's a custom implementation anyway.

I'll remove the `Store` parameter and that'll simplify things a bit.